File: 05/pt2.ts
import * as fs from 'fs';

if (process.argv.length < 3) {
  console.log('Please provide a file path');
  process.exit(1);
}

const filePath = process.argv[2]; // Assuming the file path is passed as the first command line argument
const fileString = fs.readFileSync(filePath).toString()

const chunks = (fileString.split("\n\n"))

const seedRangeMatches = chunks.shift()?.split(":")[1].trim().matchAll(/(\d+\s\d+)/ig)
const seedRanges = [...seedRangeMatches!].map(m => m[0].split(" ").map(n => parseInt(n)));

console.log(seedRanges)

const maps = chunks.map(m => m.split(":")[1].trim().split("\n").map(l => l.trim().split(" ").map(n => parseInt(n))))

const remap = (n: number, map: number[][]) => {
  // range is [dest,source,len]
  for(let i = 0; i < map.length; i++) { 
    const [dest,source,len] = map[i]
    if(n >= source && n <= source + len) {
      return dest + (n - source)
    }
  }
  return n
}

const reducingRemap = (n:number, maps: number[][][]) => {
  return maps.reduce((acc, map) => remap(acc, map), n)
}

process.stdout.write("Seeds (this is going to take a while):\n")
const resultArray:number[] = []

seedRanges.forEach(([start,len], j) => {
  for(let i = start; i < start+len; i++) {
    resultArray.push(reducingRemap(i,maps))
  }
})

process.stdout.write("\n")
console.log("the minimum of which is",Math.min(...resultArray!))

File: 05/pt2.exs
defmodule DayFiveTwo do
  def main do
    [filepath | _rest ] = System.argv()

    filestring = File.read!(filepath)

    [seedsString|mapsString] = filestring |> String.split("\n\n") |> Enum.map(&String.trim/1)
    seedMatches = Regex.scan(~r/\d+\s\d+/, seedsString)
    seedRangesAsStrings = Enum.map(seedMatches, fn e -> Enum.at(e,0) |> String.split(" ") end)
    seedRanges = Enum.map(seedRangesAsStrings, fn [a,b] -> [String.to_integer(a), String.to_integer(b)] end)
                  |> Enum.sort(fn [_,a],[_,b] -> a < b end)
    [_x|s] = seedsString |> String.split(": ") |> Enum.map(&String.trim/1)
    # seeds = s |> Enum.at(0) |> String.split(" ") |> Enum.map(&String.to_integer/1)

    maps = mapsString |> Enum.map(fn frag ->
      [_title | map] = frag |> String.split("\n") |> Enum.map(&String.trim/1)
      map |> Enum.map( fn s ->
        s |> String.split(" ") |> Enum.map(&String.to_integer/1)
      end)
    end)

    # pt1:
    # results = seeds |> Enum.map(fn seed ->
      # reducingRemap(seed,maps)
    # end)

    #pt2:
    results = Enum.flat_map(seedRanges, fn [start,len] ->
      IO.puts("start: #{start}, len: #{len}")
      Enum.map(start..(start+len), fn i ->
        reducingRemap(i,maps)
      end)
    end)


    # IO.inspect(results, charlists: :as_lists)
    IO.puts(Enum.min(results))


  end

  def remap(n, map) do
    mappings = map |> Enum.map(fn [dest,source,len] ->
      if (n >= source) && (n <= source + len) do
        dest + (n - source)
      end
    end) |> Enum.filter(&(&1 != nil))

    if Enum.count(mappings) > 0 do
      Enum.at(mappings,0)
    else
      n
    end
  end

  def reducingRemap(n,maps) do
    Enum.reduce(maps, n, fn map, acc ->
      remap(acc, map)
    end)
  end

end

DayFiveTwo.main()
